lib/native-sampler.js

```diff
@@ -64,10 +64,36 @@ NativeSampler.prototype._read = function() {
     debug('sampling %d documents from a collection with %d documents',
       this.size, count);
 
-    this.collection.aggregate(this.pipeline, options)
+    // if we need more docs than counted, use find to return all docs
+    if (count <= this.size) {
+      return this.collection.find(this.query, {
+        fields: this.fields
+      }).on('error', this.emit.bind(this, 'error'))
+        .on('data', this.push.bind(this))
+        .on('end', this.push.bind(this, null));
+    }
+
+    // if we need more than 5% of all docs, use ReservoirSampler to avoid
+    // the blocking sort stage (SERVER-22815)
+    if (count <= this.size * 20) {
+      var reservoirSampler = new ReservoirSampler(this.db, this.collectionName, this.opts);
+      return reservoirSampler
       .on('error', this.emit.bind(this, 'error'))
       .on('data', this.push.bind(this))
       .on('end', this.push.bind(this, null));
+    }
+
+    // else, use native sampler with random index walk optimization
+    this.db.command({atlasSize: 1}, function(errAtlas) {
```

Comment:
Yes, agree. And I don't feel comfortable sending an unsupported command request every time you sample, but they don't leave us much choice.

Caching is hard to do in collection-sample itself. So we'll have to expose a forced override option ("I know it doesn't support it, don't even try anymore") eventually.